docs/getting-started/package-infrastructure.md

```diff
@@ -390,7 +390,8 @@ spec:
         - fromConnectionSecretKey: username
         - fromConnectionSecretKey: password
         - fromConnectionSecretKey: endpoint
-        - name: port
+        - type: FromValue
+          name: port
```

Comment:
```console
  Warning  ComposeResources   29m (x12 over 34m)    defined/compositeresourcedefinition.apiextensions.crossplane.io  connection detail of type "FromConnectionSecretKey" key is not set
```

I noticed this change was needed because when I applied one of our doc snippets it was giving me the above warning, which means technically introducing a `type` here is a breaking behaviour change. I wonder if we're willing to tolerate it? I'm slightly tempted to, because my guess is that 'FromValue' connection details are not super common.

I believe the impact of including this change would be that XRs that used a `FromValue` connection detail would cease to successfully compose resources until their Composition was updated to be specific about its connection detail type.

Reply:
This is because now the default is `FromConnectionSecretKey`, right? I agree not many people use it but I still don't really feel comfortable breaking this in `v1`, it just doesn't feel like something important enough that's worth breaking our contract with the community. Can we work around it by having the defaulting mechanism in the code, instead of kubebuilder? Like when `type` is not given: `if value is given, the type is FromValue, else assume type is FromConnectionSecretKey`. It's hacky, but it'd be a fairly contained hack in the codebase.

Reply:
This was the behavior we had previously @muvaf, but I think there was a discussion about removing this behavior in order to make the connection detail more explicit. I think this is overall a positive change since trying to imply the type a user is intended could lead to unintended behavior.

The other breaking change here is the move to `FromValue` instead of `Value`. Although this is more consistent with the names of the other fields, I would argue it is not as valuable of a chance if we are considered breaking/updating the API version contract.

Reply:
> The other breaking change here is the move to FromValue instead of Value

I don't think this is a breaking change. The field for specifying a value is still `value:` - I just renamed one of the possible values of the `type:` field from `Value` to `FromValue`. This type field only exists in `master` (it's unreleased) so I don't think we need to consider renaming it to be a breaking change.

> Can we work around it by having the defaulting mechanism in the code, instead of kubebuilder?

I was thinking something similar. I'll add a commit to this PR.

Reply:
This is now done.